cdc/kv/shared_client.go

```diff
@@ -0,0 +1,1184 @@
+			enter := time.Now()
+			s.resolveLockCh.In() <- resolveLockTask{regionID, maxVersion, state, enter}
+		}
+	}
+
+	return rt
+}
+
+func (r *requestedTable) associateSubscriptionID(event model.RegionFeedEvent) MultiplexingEvent {
+	return MultiplexingEvent{
+		RegionFeedEvent: event,
+		SubscriptionID:  SubscriptionID(r.requestID),
+		Start:           time.Now(),
+	}
+}
+
+func (r *requestedTable) updateStaleLocks(s *SharedClient, maxVersion uint64) {
+	for {
+		old := r.staleLocks.maxVersion.Load()
+		if old >= maxVersion {
+			return
+		}
+		if r.staleLocks.maxVersion.CompareAndSwap(old, maxVersion) {
+			break
+		}
+	}
+
+	res := r.rangeLock.CheckSlowLockedRanges(r.postUpdateRegionResolvedTs)
+	log.Warn("event feed finds slow locked ranges",
+		zap.String("namespace", s.changefeed.Namespace),
+		zap.String("changefeed", s.changefeed.ID),
+		zap.String("span", r.span.String()),
+		zap.Uint64("requestID", r.requestID),
+		zap.Any("LockedRanges", res))
+}
+
+type sharedClientMetrics struct {
+	regionLockDuration      prometheus.Observer
+	regionLocateDuration    prometheus.Observer
+	regionConnectDuration   prometheus.Observer
+	batchResolvedSize       prometheus.Observer
+	lockResolveWaitDuration prometheus.Observer
+	lockResolveRunDuration  prometheus.Observer
+}
+
+func (s *SharedClient) initMetrics() {
+	s.metrics.regionLockDuration = regionConnectDuration.
+		WithLabelValues(s.changefeed.Namespace, s.changefeed.ID, "lock")
+	s.metrics.regionLocateDuration = regionConnectDuration.
+		WithLabelValues(s.changefeed.Namespace, s.changefeed.ID, "locate")
+	s.metrics.regionConnectDuration = regionConnectDuration.
+		WithLabelValues(s.changefeed.Namespace, s.changefeed.ID, "connect")
+
+	s.metrics.lockResolveWaitDuration = lockResolveDuration.
+		WithLabelValues(s.changefeed.Namespace, s.changefeed.ID, "wait")
+	s.metrics.lockResolveRunDuration = lockResolveDuration.
+		WithLabelValues(s.changefeed.Namespace, s.changefeed.ID, "run")
+
+	s.metrics.batchResolvedSize = batchResolvedEventSize.
+		WithLabelValues(s.changefeed.Namespace, s.changefeed.ID)
+}
+
+func (s *SharedClient) clearMetrics() {
+	regionConnectDuration.DeleteLabelValues(s.changefeed.Namespace, s.changefeed.ID, "lock")
+	regionConnectDuration.DeleteLabelValues(s.changefeed.Namespace, s.changefeed.ID, "locate")
+	regionConnectDuration.DeleteLabelValues(s.changefeed.Namespace, s.changefeed.ID, "connect")
+
+	lockResolveDuration.DeleteLabelValues(s.changefeed.Namespace, s.changefeed.ID, "wait")
+	lockResolveDuration.DeleteLabelValues(s.changefeed.Namespace, s.changefeed.ID, "run")
+
+	batchResolvedEventSize.DeleteLabelValues(s.changefeed.Namespace, s.changefeed.ID)
+}
+
+func hashRegionID(regionID uint64, slots int) int {
+	b := make([]byte, 8)
+	binary.LittleEndian.PutUint64(b, regionID)
+	return int(seahash.Sum64(b) % uint64(slots))
+}
+
+var (
+	// To generate a requestID in `newRequestedTable`.
+	requestIDGen atomic.Uint64
```

Comment:
```suggestion
	subscriptionIDGen atomic.Uint64
```
And please fix the comment on this field.